src/coprocessor/dag/executor/mod.rs

```diff
@@ -205,43 +210,67 @@ impl OriginCols {
         }
         Ok(values)
     }
+}
 
-    // inflate with the real value(Datum) for each columns in offsets
-    // inflate with Datum::Null for those cols not in offsets.
-    // It's used in expression since column is marked with offset
-    // in expression.
-    pub fn inflate_cols_with_offsets(
-        &self,
-        ctx: &mut EvalContext,
-        offsets: &[usize],
-    ) -> Result<Vec<Datum>> {
-        let mut res = vec![Datum::Null; self.cols.len()];
-        for offset in offsets {
-            let col = &self.cols[*offset];
-            if col.get_pk_handle() {
-                let v = util::get_pk(col, self.handle);
-                res[*offset] = v;
-            } else {
-                let col_id = col.get_column_id();
-                let value = match self.data.get(col_id) {
-                    None if col.has_default_val() => {
-                        // TODO: optimize it to decode default value only once.
-                        box_try!(table::decode_col_value(
-                            &mut col.get_default_val(),
-                            ctx,
-                            col
-                        ))
-                    }
-                    None if mysql::has_not_null_flag(col.get_flag() as u64) => {
-                        return Err(box_err!("column {} of {} is missing", col_id, self.handle));
-                    }
-                    None => Datum::Null,
-                    Some(mut bs) => box_try!(table::decode_col_value(&mut bs, ctx, col)),
-                };
-                res[*offset] = value;
-            }
+pub struct RowWithEvalContext<'a, 'b> {
```

Comment:
Is it necessary to be "RowWithEvalContext", instead of just a Row? For example, we can pass ctx in `datum_at()`?